File: GeneaGrab.Core/Helpers/Ligeo.cs
using Newtonsoft.Json.Linq;

namespace GeneaGrab.Core.Helpers
{
    public class LigeoManifest : IiifManifest<IiifSequence<LigeoCanvas>>
    {
        public LigeoManifest(string manifest) : this(JObject.Parse(manifest)) { }
        public LigeoManifest(JToken manifest) : base(manifest) {
            IsIndexable = manifest.Value<bool>("ligeoIsIndexable");
        }
        
        public bool IsIndexable { get; }
    }

    public class LigeoCanvas : IiifCanvas<IiifImage>
    {
        public LigeoCanvas(JToken canvas) : base(canvas)
        {
            Permalink = canvas.Value<string>("ligeoPermalink");
            RestrictedAccess = canvas.Value<bool>("ligeoRestrictedAccess");
            RestrictMessage = canvas.Value<string>("ligeoRestrictMessage");
            IsSearchable = canvas.Value<bool>("ligeoIsSearchable");
            Classeur = new LigeoClasseur(canvas["ligeoClasseur"]);
            MediaPath = canvas.Value<string>("ligeoMediaPath");
        }
        
        public string Permalink { get; }
        public bool RestrictedAccess { get; }
        public string RestrictMessage { get; }
        public bool IsSearchable { get; }
        public LigeoClasseur Classeur { get; }
        public string MediaPath { get; }
    }
    
    public class LigeoClasseur
    {
        public LigeoClasseur(JToken classeur)
        {
            ImageBase = classeur.Value<string>("strImageBase");
            ImageDir = classeur.Value<string>("strImageDir");
            Tag = classeur.Value<string>("curTag");
            TagNumber = classeur.Value<int>("curTagnum");
            NoticeId = classeur.Value<string>("notice_id");
            Ark = classeur.Value<string>("ark");
            UnitId = classeur.Value<string>("unitid");
            EncodedArchivalDescriptionId = classeur.Value<string>("eadid");
        }
        public string ImageDir { get; }
        public string ImageBase { get; }
        public string Tag { get; }
        public int TagNumber { get; }
        public string NoticeId { get; }
        public string Ark { get; }
        public string UnitId { get; }
        public string EncodedArchivalDescriptionId { get; }
    }
}

File: backend/providers/common/src/protocols/ligeo.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct LigeoClasseur {
    #[serde(rename = "strImageBase")]
    pub image_base: Option<String>,
    #[serde(rename = "strImageDir")]
    pub image_dir: Option<String>,
    #[serde(rename = "curTag")]
    pub tag: Option<String>,
    #[serde(rename = "curTagnum")]
    pub tag_number: Option<u32>,
    pub notice_id: Option<String>,
    pub ark: Option<String>,
    pub unitid: Option<String>,
    pub eadid: Option<String>,
}

impl TryFrom<&super::iiif::Canvas> for LigeoClasseur {
    type Error = &'static str;

    fn try_from(canvas: &super::iiif::Canvas) -> Result<Self, Self::Error> {
        let value = canvas
            .extra
            .get("ligeoClasseur")
            .ok_or("Canvas does not contain a ligeoClasseur extension")?;

        serde_json::from_value(value.clone())
            .map_err(|_| "Failed to deserialize LigeoClasseur from canvas data")
    }
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct LigeoCanvas {
    #[serde(rename = "ligeoPermalink")]
    pub permalink: Option<String>,
    #[serde(rename = "ligeoClasseur")]
    pub classeur: Option<LigeoClasseur>,
}

impl TryFrom<&super::iiif::Canvas> for LigeoCanvas {
    type Error = &'static str;

    fn try_from(canvas: &super::iiif::Canvas) -> Result<Self, Self::Error> {
        serde_json::from_value(canvas.extra.clone())
            .map_err(|_| "Failed to deserialize LigeoCanvas from canvas extra data")
    }
}
